src/_pytest/tmpdir.py

```diff
@@ -289,27 +289,26 @@ def tmp_path(
 
     del request.node.stash[tmppath_result_key]
 
-    # remove dead symlink
-    basetemp = tmp_path_factory._basetemp
-    if basetemp is None:
-        return
-    cleanup_dead_symlink(basetemp)
-
 
 def pytest_sessionfinish(session, exitstatus: Union[int, ExitCode]):
     """After each session, remove base directory if all the tests passed,
     the policy is "failed", and the basetemp is not specified by a user.
     """
     tmp_path_factory: TempPathFactory = session.config._tmp_path_factory
-    if tmp_path_factory._basetemp is None:
+    basetemp = tmp_path_factory._basetemp
+    if basetemp is None:
         return
+
+    # Remove dead symlinks.
+    cleanup_dead_symlinks(basetemp)
+
     policy = tmp_path_factory._retention_policy
     if (
         exitstatus == 0
         and policy == "failed"
         and tmp_path_factory._given_basetemp is None
     ):
```

Comment:
[Nits]
It's clearer to me to put `cleanup_dead_symlinks` after this `if` block because the whole directory would be removed in this `if` block. In that case, no need to clean up symlinks. It also leads to a little bit more of an efficiency.

Don't forget to return after the `rmtree` if you would do this.

Reply:
FYI done that slightly different, I think avoiding the early return there makes the code a bit more straightforward (but in this case I think is more a matter of taste).